PDF4Qt library support for rendering web pages to QPainter

nyalldawson opened the pull request:
This PR adds experimental support for an embedded copy of the PDF4Qt library. It's opt in via the non-default WITH_PDF4QT flag (as it requires a number of new dependancies, like lcms, fontconfig, freetype, ...) and also requires a c++20 compatible compiler. 

I've also taken a number of shortcuts with the patched in qt5 build support for this library.

If everything works, then the new QgsPdfRenderer class can be used to render a pdf document to a QPainter. 

While this gets us vectorised rendering of PDFs, the bad news is that it does **NOT** render text-as-text, rather as non-text paths. So if we move to this approach for rendering HTML to painter (via webengine pdf exports), then we'll lose text-as-text support... but it's still a LOT nicer than forced low-resolution rasterised HTML rendering...

In short, the cluster**** which is Qt's web support continues to bite. :shrug:

nyalldawson committed:
Create QgsWebEnginePage wrapper class around QWebEnginePage

Like QgsWebPage this provides additional functional over the
base Qt class to simplify tasks required in QGIS.

nyalldawson committed:
Some cmake tweaks for PDF4QT

nyalldawson committed:
Add basic find lcms2 module

nyalldawson committed:
Fix library include for lcms2

nyalldawson committed:
Build PDF4QT along with core

nyalldawson committed:
Implement downstream patch to add Qt5 support

nyalldawson committed:
Add fontconfig

nyalldawson committed:
Make building PDF4Qt optional, and opt-in

There's some added dependancies required for building this, so
let's play it safe for now...

nyalldawson committed:
Shell for QgsPdfRenderer class

nyalldawson committed:
Lots of shortcuts to get a PDF -> QPainter renderer working

nyalldawson committed:
Fix build without PDF4Qt

nyalldawson committed:
Test building on qt6 ci with PDF4Qt

troopa81 commented:
What PDF4QT version did you copy as external? Could be nice to keep a track of this information somewhere (in the CMakeLists.txt where we include it for instance), it could be useful when we would like to upgrade it some days.

> I've also taken a number of shortcuts with the patched in qt5 build support for this library.

The patches are only required for qt5 ? If yes, is there an interest in keeping those patches and why not making PDF4QT a library used only when we build with Qt6 ?

> While this gets us vectorised rendering of PDFs, the bad news is that it does NOT render text-as-text, rather as non-text paths. So if we move to this approach for rendering HTML to painter (via webengine pdf exports), then we'll lose text-as-text support...

Is it a well-known limitation ? Is it possible to fix it upstream?

I imagine this is only the API PR and more would come? Because I see no integration with current code

nyalldawson commented:
@troopa81 

> What PDF4QT version did you copy as external? Could be nice to keep a track of this information somewhere (in the CMakeLists.txt where we include it for instance)

Can do -- I'll put the upstream sha in a text file in the external dir.

> The patches are only required for qt5 ? If yes, is there an interest in keeping those patches and why not making PDF4QT a library used only when we build with Qt6 ?

They are quite trivial patches, and I'd rather keep the functionality diff small between qt5/6 builds. It makes it easier to develop scripts/plugins which work consistently everywhere.

> Is it a well-known limitation ? Is it possible to fix it upstream?

I don't think it's impossible at least 😂. I'm not familiar myself with pdf internals to know if there's anything blocking this. I'll check with upstream and see their thoughts.

> I imagine this is only the API PR and more would come? Because I see no integration with current code

It's just a draft of the embedding for now. If this looks acceptable, then later there'll be a class which handles html-> painter rendering via webengine and a patch for the layout item.

rouault commented:
> > 
> 
> I don't think it's impossible at least 😂. I'm not familiar myself with pdf internals to know if there's anything blocking this. I'll check with upstream and see their thoughts.

if that might help, GDAL's PDF very basic text rendering (limited to ASCII or Latin1) as text: https://github.com/OSGeo/gdal/blob/master/frmts/pdf/pdfcreatecopy.cpp#L3205. The key PDF instructions are: "BT (some text)Tj ET"

rouault commented on external/PDF4QT/3rdparty_licenses/LittleCMS_COPYING.txt:
```diff
@@ -0,0 +1,8 @@
+Little CMS
```
that file shouldn't be necessary as we don't embed lcms source code. Same for other files in that directory

nyalldawson replied:
Fixed, thanks

andreasneumann commented:
@nyalldawson - thank you for working on this!

Question: now that we work with the Qt developers to integrate CMYK into Qt, specifically for PDF and TIFF. This would probably mean for users that they will have to choose: either get CMYK PDF (but with web/HTML converted to raster), or to get web/HTML content as vector but no CMYK - right? Or do I misinterpret the situation?

The pdf4qt - would it support CMYK?

For most users it is probably a non-issue, because they will continue to use RGB and won't go to print. But the ones who will use CMYK will probably the first users who will complain that their web/html content is rasterized ...

nyalldawson commented:
@andreasneumann I don't think html really supports cmyk, does it? If not then we'll always be going rgb html -> RGB pdf -> convert to CMYK during layout export anyway...

andreasneumann commented:
> @andreasneumann I don't think html really supports cmyk, does it? If not then we'll always be going rgb html -> RGB pdf -> convert to CMYK during layout export anyway...

yes, HTML doesn't support CMYK. So that part would have to be converted from RGB to CMYK.

How would this solution with the PDF4Qt library work? Would only the HTML/Web layout items go through this library and the rest would of the layout (map, map items) would take the other route directly through the QtPDF renderer?

nyalldawson commented:
@andreasneumann 

> How would this solution with the PDF4Qt library work? Would only the HTML/Web layout items go through this library and the rest of the layout (map, map related items, legend, etc.) would take the other route directly through the QtPDF renderer, thus taking advantage of CMYK support?

Oh, it's not anywhere NEAR that sensible 😂

Html items alone will go from webengine (RGB) -> pdf (RGB) > pdf4qt (rgb) > Qpainter (RGB onto either RGB or CMYK surface) to render back in the layout as vector and THEN be re-exported via qt's pdf exporter along with the rest of the layout (in RGB or CMYK) 🤪

troopa81 commented:
> Html items alone will go from webengine (RGB) -> pdf (RGB) > pdf4qt (rgb) > Qpainter (RGB onto either RGB or CMYK surface)

It would not be possible to render onto a CYMK image, it would only be possible to render in CYMK directly in PDF. If I understand correctly the HTML rendering will be rendered in RGB while the symbology could be rendered as CYMK vector elements, all mixed in the same PDF. Am I right?

nyalldawson commented:
@troopa81 
>  would not be possible to render onto a CYMK image, it would only be possible to render in CYMK directly in PDF.

No, that's incorrect. It IS possible to render RGB content onto a CMYK surface.

>  If I understand correctly the HTML rendering will be rendered in RGB while the symbology could be rendered as CYMK vector elements, all mixed in the same PDF. Am I right?

Think of it in the same way as if you had a svg item in the project pointing to a RGB svg -- it can still be rendered when using CMYK colour space, it just gets a naive RGB to CMYK conversion at render time. Or if some elements in your projects symbology have colors set to RGB and not CMYK.

Same behaviour as embedding an RGB image into a CMYK drawing in eg Adobe Illustrator.

You can use RGB colors in a CMYK drawing/project, they just converted at render/export time. Similarly you can use CMYK colors in a RGB drawing and they'll also get converted.

troopa81 commented:
> Think of it in the same way as if you had a svg item in the project pointing to a RGB svg -- it can still be rendered when using CMYK colour space, it just gets a naive RGB to CMYK conversion at render time. Or if some elements in your projects symbology have colors set to RGB and not CMYK.

OK, it answers my concerns.

> No, that's incorrect. It IS possible to render RGB content onto a CMYK surface.

How is that possible? Or maybe I'am missing something. It is not possible to even defined a QImage with CYMK content for now. Once KDAB work is complete, it would be possible to:
- load a CYMK image
- paint CYMK into a PDF into native CYMK color (without RGB to CMYK conversion)

But it would not be possible to do [paint in a CYMK image](https://github.com/dangelog/qtbase-cmyk/blob/cmyk/src/gui/painting/qpainter.cpp#L1773).

As I understand it, this is fine because we would render everything directly in CMYK colorspace in PDF, except for the HTML part that would be converted from RGB.

nyalldawson commented:
> How is that possible? Or maybe I'am missing something. It is not possible to even defined a QImage with CYMK content for now. Once KDAB work is complete, it would be possible to:

> load a CYMK image
paint CYMK into a PDF into native CYMK color (without RGB to CMYK conversion)
But it would not be possible to do paint in a CYMK image.

Well that's an implementation detail. It's still theoretically possible to render content from either colour space into the other by doing an on the fly conversion.

> As I understand it, this is fine because we would render everything directly in CMYK colorspace in PDF, except for the HTML part that would be converted from RGB

And picture items, and labels/tables/scale bars/map layers/north arrows/... Etc where ANY colour used is non CMYK. This isn't a scenario unique to html rendering.  Maybe we can move the discussion to a CMYK qep or something?

nyalldawson committed:
Fix deprecated implicit capture

nyalldawson committed:
Cleanup API

nyalldawson committed:
Fix build

nyalldawson committed:
Fix warning

nyalldawson committed:
Add unit test for pdf rendering

nyalldawson committed:
Add QgsWebEnginePage::render( QPainter* ) method

Requires a PDF4Qt enabled build. This method exports the QWebEnginePage
to a PDF file, so that we get nicely vectorised objects and
text-as-text. It's the ONLY supported way in Qt API to do this, every
other method of rendering QWebEnginePage forces rasterisation of
the content.

We then use the QgsPdfRenderer class to render the temporary PDF
content to a QPainter device, which keeps vector objects as vectors.

nyalldawson committed:
Add method to determine actual document size

nyalldawson committed:
Fix page height during render

nyalldawson committed:
Add test for rendering html with text

nyalldawson commented:
This is ready for review/merge now. We've now got two new classes:

- QgsPdfRenderer. Enabled when WITH_PDF4QT cmake flag is on. This class handles rendering of PDF documents to QPainter surfaces, keeping vector objects as vectors (using the lib PDF4Qt backend). 
- QgsWebEnginePage: A wrapper around the QWebEnginePage class, adding extra functionality. Enabled only when the QGIS build has webengine support. When WITH_PDF4QT is ALSO set, then there's a convenience method QgsWebEnginePage::render which permits directly rendering HTML contents to a QPainter destination, keeping vector as vector and using the modern webengine browser engine to handle the actual HTML /javascript rendering.

nyalldawson committed:
Disable webengine sandboxing on ci

rouault commented on tests/src/python/test_qgspdfrenderer.py:
```diff
@@ -0,0 +1,97 @@
+                         QRectF(0.0, 0.0, 420.0, 595.0))
+        self.assertEqual(renderer.pageMediaBox(2), QRectF())
+
+    def test_pdf_render(self):
```
maybe add a test by providing something that is not a PDF to QgsPdfRenderer ?

nyalldawson replied:
Sure, done!

nyalldawson committed:
Fix qt5 build with PDF support

nyalldawson committed:
Test opening a file which isn't a PDF
